src/libopensc/card-coolkey.c

```diff
@@ -1757,6 +1757,9 @@ static int coolkey_rsa_op(sc_card_t *card,
 		u8 len_buf[2];
 		size_t out_length;
 
+		/* Free card response -- nothing useful -- result is in export object */
+		free(*crypt_out_p);
```

Comment:
Are you sure that this is correct? It looks like there is some case where this calls free on a buffer on the heap:
https://github.com/OpenSC/OpenSC/blob/0d958c1f7d8ea34e3e127bf560dcbf963b9147a9/src/libopensc/card-coolkey.c#L1702-L1703

Reply:
The code is very confusing, but most important is that `crypt_in_len = datalen;` and therefore the conditions `if (crypt_in_len > MAX_COMPUTE_BUF) {` and `if (datalen > MAX_COMPUTE_BUF) {` are equal. In the first one we are setting

https://github.com/OpenSC/OpenSC/blob/0d958c1f7d8ea34e3e127bf560dcbf963b9147a9/src/libopensc/card-coolkey.c#L1725

which keeps the io function to allocate the memory. This probably does not happen with real cards, but fuzzer can return some mess somethimes.

Reply:
`crypt_out_p` is only reset `if (crypt_in_len > MAX_COMPUTE_BUF)`, so there seems to be the other case where it is *not* set to `NULL`. Would it be possible to clearify the code, e.g. by using different variables?

Reply:
I will try to improve that part.

Reply:
Rewritten in a8101c26. Does it look better for you now?

Reply:
The final version is in 6b66c43 as the initial had some issues to build. Looks fine now.